### ISSUE TITLE
Cannot list vertices after I create one

### ISSUE THREAD

**Issue by ihoro** [lambdazen/bitsy#33]
Hi,

I've combined the latest gremlin-server (v3.4.9) with the latest bitsy (v3.4.0) and trying to create a vertex and read it after. But I do not get a result for the listing. You can find dockerized test [here](https://github.com/ihoro/bitsy-playground/tree/master/issue01-cannot-list-after-create).

This is how the server starts:
```
$ docker-compose up db
Creating network "issue01-cannot-list-after-create_net" with the default driver
Creating issue01-cannot-list-after-create_db_1 ... done
Attaching to issue01-cannot-list-after-create_db_1
db_1    | [INFO] GremlinServer - 3.4.9
db_1    |          \,,,/
db_1    |          (o o)
db_1    | -----oOOo-(3)-oOOo-----
db_1    |
db_1    | [INFO] GremlinServer - Configuring Gremlin Server from conf/gremlin-server-bitsy.yaml
db_1    | [INFO] MetricManager - Configured Metrics Slf4jReporter configured with interval=180000ms and loggerName=org.apache.tinkerpop.gremlin.server.Settings$Slf4jReporterMetrics
db_1    | [INFO] FileBackedMemoryGraphStore - Starting graph FileBackedMemoryGraphStore-1(path = /bitsy-data)
db_1    | [INFO] DefaultGraphManager - Graph [graph] was successfully configured via [conf/bitsy.properties].
db_1    | [INFO] ServerGremlinExecutor - Initialized Gremlin thread pool.  Threads in pool named with pattern gremlin-*
db_1    | [INFO] ServerGremlinExecutor - Initialized GremlinExecutor and preparing GremlinScriptEngines instances.
db_1    | [INFO] ServerGremlinExecutor - Initialized gremlin-groovy GremlinScriptEngine and registered metrics
db_1    | [INFO] ServerGremlinExecutor - A GraphTraversalSource is now bound to [g] with graphtraversalsource[bitsygraph[/bitsy-data], standard]
db_1    | [INFO] OpLoader - Adding the standard OpProcessor.
db_1    | [INFO] OpLoader - Adding the session OpProcessor.
db_1    | [INFO] OpLoader - Adding the traversal OpProcessor.
db_1    | [INFO] TraversalOpProcessor - Initialized cache for TraversalOpProcessor with size 1000 and expiration time of 600000 ms
db_1    | [INFO] GremlinServer - Executing start up LifeCycleHook
db_1    | [INFO] Logger$info - Executed once at startup of Gremlin Server.
db_1    | [INFO] GremlinServer - idleConnectionTimeout was set to 0 which resolves to 0 seconds when configuring this value - this feature will be disabled
db_1    | [INFO] GremlinServer - keepAliveInterval was set to 0 which resolves to 0 seconds when configuring this value - this feature will be disabled
db_1    | [WARN] AbstractChannelizer - The org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0 serialization class is deprecated.
db_1    | [INFO] AbstractChannelizer - Configured application/vnd.gremlin-v3.0+gryo with org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0
db_1    | [WARN] AbstractChannelizer - The org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0 serialization class is deprecated.
db_1    | [INFO] AbstractChannelizer - Configured application/vnd.gremlin-v3.0+gryo-stringd with org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0
db_1    | [INFO] AbstractChannelizer - Configured application/vnd.gremlin-v3.0+json with org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0
db_1    | [INFO] AbstractChannelizer - Configured application/json with org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0
db_1    | [INFO] GremlinServer$1 - Gremlin Server configured with worker thread pool of 1, gremlin pool of 4 and boss thread pool of 1.
db_1    | [INFO] GremlinServer$1 - Channel started at port 8182.
```
And these are my tryouts:
```
$ docker-compose run --rm test list
Testing...
[]
$ docker-compose run --rm test list
Testing...
[]
$ docker-compose run --rm test create
Testing...
{
  value: Vertex {
    id: '2e9f5ad8-184e-408b-8282-128b29ef3f0e',
    label: 'Person',
    properties: undefined
  },
  done: false
}
$ docker-compose run --rm test list
Testing...
# and it just hangs on this step, and gremlin-server's log is silent - no errors there
```

What do you think? I hope that I simply do something wrong.

**Comment by lambdazen** [comment]
Hi @ihoro,

Hanging isn't good. Would you be able to get a stack trace for the hanging process? https://www.baeldung.com/java-thread-dump

This could be an issue with
- the docker environment
- multiple processes accessing the database files
- gremlin server losing connectivity (or something along those lines)
- missing commit (unlikely) -- you can add a graph.commit() just to be sure

The stack trace should help debug any of these possibilities. 

Regards,

Sridhar.

**Comment by ihoro** [comment]
👋 @lambdazen ,

This is what I've found:

Before any gremlin query, i.e. after start of the server:
```
2020-12-23 13:45:44
Full thread dump OpenJDK 64-Bit Server VM (25.212-b04 mixed mode):

"Attach Listener" #18 daemon prio=9 os_prio=0 tid=0x000056061fcaf000 nid=0x77 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"DestroyJavaVM" #17 prio=5 os_prio=0 tid=0x0000560620290800 nid=0x16 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"gremlin-server-boss-1" #16 prio=5 os_prio=0 tid=0x000056061f5b7800 nid=0x2f runnable [0x00007fbe80240000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000006c016b4f8> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x00000006c016c608> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000006c016c520> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:68)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:805)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-1" #15 prio=5 os_prio=0 tid=0x000056062013b000 nid=0x28 waiting on condition [0x00007fbe80342000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"VEReorg-1" #14 daemon prio=5 os_prio=0 tid=0x000056061ff12000 nid=0x27 in Object.wait() [0x00007fbe80843000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:40)
	- locked <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)

"TxFlusher-1" #13 daemon prio=5 os_prio=0 tid=0x000056062001f000 nid=0x26 in Object.wait() [0x00007fbe80986000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c0142e20> (a com.lambdazen.bitsy.store.TxLogFlushPotential)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c0142e20> (a com.lambdazen.bitsy.store.TxLogFlushPotential)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:43)

"MemToTxLogWriter-1" #12 daemon prio=5 os_prio=0 tid=0x000056061fff3000 nid=0x25 in Object.wait() [0x00007fbe80a87000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013c318> (a com.lambdazen.bitsy.store.FileBackedMemoryGraphStore$1)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c013c318> (a com.lambdazen.bitsy.store.FileBackedMemoryGraphStore$1)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:43)

"metrics-logger-reporter-thread-1" #10 daemon prio=5 os_prio=0 tid=0x000056061fb4d800 nid=0x24 waiting on condition [0x00007fbe844f5000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0142fe8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"Service Thread" #8 daemon prio=9 os_prio=0 tid=0x000056061f230800 nid=0x22 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread2" #7 daemon prio=9 os_prio=0 tid=0x000056061f21d800 nid=0x21 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #6 daemon prio=9 os_prio=0 tid=0x000056061f1c7000 nid=0x20 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 tid=0x000056061f1ba000 nid=0x1f waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x000056061f1b7800 nid=0x1e runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x000056061f183800 nid=0x1d in Object.wait() [0x00007fbe84e05000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013cc30> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	- locked <0x00000006c013cc30> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x000056061f180800 nid=0x1c in Object.wait() [0x00007fbe84f06000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013cde8> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x00000006c013cde8> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"VM Thread" os_prio=0 tid=0x000056061f176000 nid=0x1b runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x000056061f046800 nid=0x17 runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x000056061f048000 nid=0x18 runnable 

"GC task thread#2 (ParallelGC)" os_prio=0 tid=0x000056061f04a000 nid=0x19 runnable 

"GC task thread#3 (ParallelGC)" os_prio=0 tid=0x000056061f04b800 nid=0x1a runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x000056061f232800 nid=0x23 waiting on condition 

JNI global references: 646
```
Right after vertex creation:
```
2020-12-23 13:46:26
Full thread dump OpenJDK 64-Bit Server VM (25.212-b04 mixed mode):

"gremlin-server-exec-4" #22 prio=5 os_prio=0 tid=0x0000560620d05000 nid=0x8e waiting on condition [0x00007fbe7dd52000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-3" #21 prio=5 os_prio=0 tid=0x00005606239ea800 nid=0x8d waiting on condition [0x00007fbe7f134000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-2" #20 prio=5 os_prio=0 tid=0x000056061f38d000 nid=0x8c waiting on condition [0x00007fbe7d6b7000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-worker-1" #19 prio=5 os_prio=0 tid=0x000056061f42b800 nid=0x8b runnable [0x00007fbe7f435000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000006c0252348> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x00000006c0253448> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000006c0253370> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:62)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:809)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)

"Attach Listener" #18 daemon prio=9 os_prio=0 tid=0x000056061fcaf000 nid=0x77 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"DestroyJavaVM" #17 prio=5 os_prio=0 tid=0x0000560620290800 nid=0x16 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"gremlin-server-boss-1" #16 prio=5 os_prio=0 tid=0x000056061f5b7800 nid=0x2f runnable [0x00007fbe80240000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000006c016b4f8> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x00000006c016c608> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000006c016c520> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:68)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:805)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-1" #15 prio=5 os_prio=0 tid=0x000056062013b000 nid=0x28 waiting on condition [0x00007fbe80342000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"VEReorg-1" #14 daemon prio=5 os_prio=0 tid=0x000056061ff12000 nid=0x27 in Object.wait() [0x00007fbe80843000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:40)
	- locked <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)

"TxFlusher-1" #13 daemon prio=5 os_prio=0 tid=0x000056062001f000 nid=0x26 in Object.wait() [0x00007fbe80986000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c0142e20> (a com.lambdazen.bitsy.store.TxLogFlushPotential)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c0142e20> (a com.lambdazen.bitsy.store.TxLogFlushPotential)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:43)

"MemToTxLogWriter-1" #12 daemon prio=5 os_prio=0 tid=0x000056061fff3000 nid=0x25 in Object.wait() [0x00007fbe80a87000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013c318> (a com.lambdazen.bitsy.store.FileBackedMemoryGraphStore$1)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c013c318> (a com.lambdazen.bitsy.store.FileBackedMemoryGraphStore$1)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:43)

"metrics-logger-reporter-thread-1" #10 daemon prio=5 os_prio=0 tid=0x000056061fb4d800 nid=0x24 waiting on condition [0x00007fbe844f5000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0142fe8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"Service Thread" #8 daemon prio=9 os_prio=0 tid=0x000056061f230800 nid=0x22 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread2" #7 daemon prio=9 os_prio=0 tid=0x000056061f21d800 nid=0x21 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #6 daemon prio=9 os_prio=0 tid=0x000056061f1c7000 nid=0x20 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 tid=0x000056061f1ba000 nid=0x1f waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x000056061f1b7800 nid=0x1e runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x000056061f183800 nid=0x1d in Object.wait() [0x00007fbe84e05000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013cc30> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	- locked <0x00000006c013cc30> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x000056061f180800 nid=0x1c in Object.wait() [0x00007fbe84f06000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013cde8> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x00000006c013cde8> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"VM Thread" os_prio=0 tid=0x000056061f176000 nid=0x1b runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x000056061f046800 nid=0x17 runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x000056061f048000 nid=0x18 runnable 

"GC task thread#2 (ParallelGC)" os_prio=0 tid=0x000056061f04a000 nid=0x19 runnable 

"GC task thread#3 (ParallelGC)" os_prio=0 tid=0x000056061f04b800 nid=0x1a runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x000056061f232800 nid=0x23 waiting on condition 

JNI global references: 716
```
And after we issue listing:
```
2020-12-23 13:47:20
Full thread dump OpenJDK 64-Bit Server VM (25.212-b04 mixed mode):

"gremlin-server-exec-4" #22 prio=5 os_prio=0 tid=0x0000560620d05000 nid=0x8e waiting on condition [0x00007fbe7dd52000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-3" #21 prio=5 os_prio=0 tid=0x00005606239ea800 nid=0x8d waiting on condition [0x00007fbe7f134000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-2" #20 prio=5 os_prio=0 tid=0x000056061f38d000 nid=0x8c waiting on condition [0x00007fbe7d6b7000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-worker-1" #19 prio=5 os_prio=0 tid=0x000056061f42b800 nid=0x8b runnable [0x00007fbe7f434000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000006c0252348> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x00000006c0253448> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000006c0253370> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:68)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:805)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)

"Attach Listener" #18 daemon prio=9 os_prio=0 tid=0x000056061fcaf000 nid=0x77 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"DestroyJavaVM" #17 prio=5 os_prio=0 tid=0x0000560620290800 nid=0x16 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"gremlin-server-boss-1" #16 prio=5 os_prio=0 tid=0x000056061f5b7800 nid=0x2f runnable [0x00007fbe80240000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000006c016b4f8> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x00000006c016c608> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000006c016c520> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:68)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:805)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)

"gremlin-server-exec-1" #15 prio=5 os_prio=0 tid=0x000056062013b000 nid=0x28 waiting on condition [0x00007fbe80342000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0003920> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"VEReorg-1" #14 daemon prio=5 os_prio=0 tid=0x000056061ff12000 nid=0x27 in Object.wait() [0x00007fbe80843000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:40)
	- locked <0x00000006c013c138> (a com.lambdazen.bitsy.store.VEObsolescencePotential)

"TxFlusher-1" #13 daemon prio=5 os_prio=0 tid=0x000056062001f000 nid=0x26 in Object.wait() [0x00007fbe80986000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c0142e20> (a com.lambdazen.bitsy.store.TxLogFlushPotential)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c0142e20> (a com.lambdazen.bitsy.store.TxLogFlushPotential)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:43)

"MemToTxLogWriter-1" #12 daemon prio=5 os_prio=0 tid=0x000056061fff3000 nid=0x25 in Object.wait() [0x00007fbe80a87000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013c318> (a com.lambdazen.bitsy.store.FileBackedMemoryGraphStore$1)
	at java.lang.Object.wait(Object.java:502)
	at com.lambdazen.bitsy.util.DoubleBuffer.getBufferToFlush(DoubleBuffer.java:110)
	- locked <0x00000006c013c318> (a com.lambdazen.bitsy.store.FileBackedMemoryGraphStore$1)
	at com.lambdazen.bitsy.util.DoubleBufferThread.doFlush(DoubleBufferThread.java:53)
	at com.lambdazen.bitsy.util.DoubleBufferThread.run(DoubleBufferThread.java:43)

"metrics-logger-reporter-thread-1" #10 daemon prio=5 os_prio=0 tid=0x000056061fb4d800 nid=0x24 waiting on condition [0x00007fbe844f5000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006c0142fe8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"Service Thread" #8 daemon prio=9 os_prio=0 tid=0x000056061f230800 nid=0x22 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread2" #7 daemon prio=9 os_prio=0 tid=0x000056061f21d800 nid=0x21 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #6 daemon prio=9 os_prio=0 tid=0x000056061f1c7000 nid=0x20 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 tid=0x000056061f1ba000 nid=0x1f waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x000056061f1b7800 nid=0x1e runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x000056061f183800 nid=0x1d in Object.wait() [0x00007fbe84e05000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013cc30> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	- locked <0x00000006c013cc30> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x000056061f180800 nid=0x1c in Object.wait() [0x00007fbe84f06000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c013cde8> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x00000006c013cde8> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"VM Thread" os_prio=0 tid=0x000056061f176000 nid=0x1b runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x000056061f046800 nid=0x17 runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x000056061f048000 nid=0x18 runnable 

"GC task thread#2 (ParallelGC)" os_prio=0 tid=0x000056061f04a000 nid=0x19 runnable 

"GC task thread#3 (ParallelGC)" os_prio=0 tid=0x000056061f04b800 nid=0x1a runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x000056061f232800 nid=0x23 waiting on condition 

JNI global references: 716
```

I hope it helps. Let me know if it needs another experiment to run or more details.

**Comment by lambdazen** [comment]
The threads "VEReorg-1", "TxFlusher-1" and "MemToTxLogWriter-1" are background threads and seem to be in the same state before and after listing causes the system to hang. 

You can look at files in the bitsy database path. These are text files with one JSON entry and hashcode per line. If you can make sure the files append new lines as you add vertices and edges. If so, that should fully eliminate the background threads as suspects. 

My suspicion is one of these: 
- Some fatal unhandled exception such as "MethodNotFound" is causing the vertex listing to fail. You can try other query methods like vertex-by-id or get-edges to see if they work. I think Gremlin uses slf4j and logback. You can set it to debug mode and see if any exceptions are happening there. 
- Requests/responses are getting lost somewhere in Gremlin, Netty or Docker: This is worth pursuing only if you are sure no exceptions occur along the way. 

There is second and third thread dumps show a minor diff in "gremlin-server-worker-1". 
```
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:68)
```

I searched for the second line and found some hits like this: https://bugs.openjdk.java.net/browse/JDK-8155861

This can't be the root cause though unless you are on a really old JDK version. 

Also, are you running the client & server in separate JVMs? Can you see what the thread state is at the client? 

Regards,

Sridhar.

**Comment by ihoro** [comment]
Yes, I always see that he saves it to fs, since the very first tests:
```
/bitsy-data # grep Igor *
vA.txt:V={"id":"81eee3e7-fdeb-43e7-820d-0566154f9a90","v":1,"s":"M","p":{"name":"Igor"},"l":"Person"}#3ff5e53c
/bitsy-data #
```

**Comment by lambdazen** [comment]
That's good. Do you see any exceptions in the log files? 

Does the hanging issue only happen when you list all vertices? Or even if you get a vertex by ID (such as "81eee3e7-fdeb-43e7-820d-0566154f9a90")?

**Comment by ihoro** [comment]
I see that gremlin-server docker image is based on Java 1.8. I've switched it to 9 and now I see an exception upon vertex creation:
```
db_1    | [INFO] GremlinServer - 3.4.9
db_1    |          \,,,/
db_1    |          (o o)
db_1    | -----oOOo-(3)-oOOo-----
db_1    |
db_1    | [INFO] GremlinServer - Configuring Gremlin Server from conf/gremlin-server-bitsy.yaml
db_1    | [INFO] MetricManager - Configured Metrics Slf4jReporter configured with interval=180000ms and loggerName=org.apache.tinkerpop.gremlin.server.Settings$Slf4jReporterMetrics
db_1    | [INFO] FileBackedMemoryGraphStore - Starting graph FileBackedMemoryGraphStore-1(path = /bitsy-data)
db_1    | [INFO] DefaultGraphManager - Graph [graph] was successfully configured via [conf/bitsy.properties].
db_1    | [INFO] ServerGremlinExecutor - Initialized Gremlin thread pool.  Threads in pool named with pattern gremlin-*
db_1    | [INFO] ServerGremlinExecutor - Initialized GremlinExecutor and preparing GremlinScriptEngines instances.
db_1    | WARNING: An illegal reflective access operation has occurred
db_1    | WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/opt/gremlin-server/lib/groovy-2.5.11-indy.jar) to method java.lang.Object.finalize()
db_1    | WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
db_1    | WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
db_1    | WARNING: All illegal access operations will be denied in a future release
db_1    | [INFO] ServerGremlinExecutor - Initialized gremlin-groovy GremlinScriptEngine and registered metrics
db_1    | [INFO] ServerGremlinExecutor - A GraphTraversalSource is now bound to [g] with graphtraversalsource[bitsygraph[/bitsy-data], standard]
db_1    | [INFO] OpLoader - Adding the standard OpProcessor.
db_1    | [INFO] OpLoader - Adding the session OpProcessor.
db_1    | [INFO] OpLoader - Adding the traversal OpProcessor.
db_1    | [INFO] TraversalOpProcessor - Initialized cache for TraversalOpProcessor with size 1000 and expiration time of 600000 ms
db_1    | [INFO] GremlinServer - Executing start up LifeCycleHook
db_1    | [INFO] Logger$info - Executed once at startup of Gremlin Server.
db_1    | [INFO] GremlinServer - idleConnectionTimeout was set to 0 which resolves to 0 seconds when configuring this value - this feature will be disabled
db_1    | [INFO] GremlinServer - keepAliveInterval was set to 0 which resolves to 0 seconds when configuring this value - this feature will be disabled
db_1    | [WARN] AbstractChannelizer - The org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0 serialization class is deprecated.
db_1    | [INFO] AbstractChannelizer - Configured application/vnd.gremlin-v3.0+gryo with org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0
db_1    | [WARN] AbstractChannelizer - The org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0 serialization class is deprecated.
db_1    | [INFO] AbstractChannelizer - Configured application/vnd.gremlin-v3.0+gryo-stringd with org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0
db_1    | [INFO] AbstractChannelizer - Configured application/vnd.gremlin-v3.0+json with org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0
db_1    | [INFO] AbstractChannelizer - Configured application/json with org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0
db_1    | [INFO] GremlinServer$1 - Gremlin Server configured with worker thread pool of 1, gremlin pool of 4 and boss thread pool of 1.
db_1    | [INFO] GremlinServer$1 - Channel started at port 8182.
db_1    | java.lang.IllegalStateException
db_1    |       at com.google.common.base.Preconditions.checkState(Preconditions.java:133)
db_1    |       at com.nesscomputing.uuid.NessUUID.<clinit>(NessUUID.java:204)
db_1    |       at com.lambdazen.bitsy.UUID.toString(UUID.java:90)
db_1    |       at com.lambdazen.bitsy.BitsyGraphSONModule$UUIDSerializer.serializeWithType(BitsyGraphSONModule.java:100)
db_1    |       at com.lambdazen.bitsy.BitsyGraphSONModule$UUIDSerializer.serializeWithType(BitsyGraphSONModule.java:76)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3058)
db_1    |       at org.apache.tinkerpop.shaded.jackson.core.base.GeneratorBase.writeObject(GeneratorBase.java:388)
db_1    |       at org.apache.tinkerpop.shaded.jackson.core.JsonGenerator.writeObjectField(JsonGenerator.java:1838)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.GraphSONSerializersV3d0$VertexJacksonSerializer.serialize(GraphSONSerializersV3d0.java:102)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.GraphSONSerializersV3d0$VertexJacksonSerializer.serialize(GraphSONSerializersV3d0.java:88)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.std.StdScalarSerializer.serializeWithType(StdScalarSerializer.java:46)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3058)
db_1    |       at org.apache.tinkerpop.shaded.jackson.core.base.GeneratorBase.writeObject(GeneratorBase.java:388)
db_1    |       at org.apache.tinkerpop.shaded.jackson.core.JsonGenerator.writeObjectField(JsonGenerator.java:1838)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.TraversalSerializersV3d0$TraverserJacksonSerializer.serialize(TraversalSerializersV3d0.java:237)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.TraversalSerializersV3d0$TraverserJacksonSerializer.serialize(TraversalSerializersV3d0.java:226)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.std.StdScalarSerializer.serializeWithType(StdScalarSerializer.java:46)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3058)
db_1    |       at org.apache.tinkerpop.shaded.jackson.core.base.GeneratorBase.writeObject(GeneratorBase.java:388)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.JavaUtilSerializersV3d0$ListJacksonSerializer.serialize(JavaUtilSerializersV3d0.java:127)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.JavaUtilSerializersV3d0$ListJacksonSerializer.serializeWithType(JavaUtilSerializersV3d0.java:135)
db_1    |       at org.apache.tinkerpop.gremlin.structure.io.graphson.JavaUtilSerializersV3d0$ListJacksonSerializer.serializeWithType(JavaUtilSerializersV3d0.java:118)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
db_1    |       at org.apache.tinkerpop.gremlin.driver.ser.AbstractGraphSONMessageSerializerV2d0$ResponseMessageSerializer.ser(AbstractGraphSONMessageSerializerV2d0.java:276)
db_1    |       at org.apache.tinkerpop.gremlin.driver.ser.AbstractGraphSONMessageSerializerV2d0$ResponseMessageSerializer.serializeWithType(AbstractGraphSONMessageSerializerV2d0.java:250)
db_1    |       at org.apache.tinkerpop.gremlin.driver.ser.AbstractGraphSONMessageSerializerV2d0$ResponseMessageSerializer.serializeWithType(AbstractGraphSONMessageSerializerV2d0.java:235)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4409)
db_1    |       at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper.writeValueAsBytes(ObjectMapper.java:3687)
db_1    |       at org.apache.tinkerpop.gremlin.driver.ser.AbstractGraphSONMessageSerializerV2d0.serializeResponseAsBinary(AbstractGraphSONMessageSerializerV2d0.java:85)
db_1    |       at org.apache.tinkerpop.gremlin.server.op.AbstractOpProcessor.makeFrame(AbstractOpProcessor.java:344)
db_1    |       at org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor.handleIterator(TraversalOpProcessor.java:587)
db_1    |       at org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor.lambda$iterateBytecodeTraversal$4(TraversalOpProcessor.java:411)
db_1    |       at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
db_1    |       at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
db_1    |       at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
db_1    |       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
db_1    |       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
db_1    |       at java.base/java.lang.Thread.run(Thread.java:844)
```
If I run listing after that it hangs as before.

**Comment by lambdazen** [comment]
This might be related to PR #32

I have released version 3.4.1 after merging this PR. Please let me know if that unblocks you.

**Comment by ihoro** [comment]
I've applied several quick tests and I experience the same hanging. I've checked it with java8, java9, and java11. See updated test stand [here](https://github.com/ihoro/bitsy-playground/tree/master/issue01-cannot-list-after-create).

**Comment by giovibal** [comment]
> at com.nesscomputing.uuid.NessUUID.<clinit>(NessUUID.java:204)

This line in stacktrace show the use of NessUUID, maybe runtime is not in sync with last release ?

**Comment by giovibal** [comment]
> > at com.nesscomputing.uuid.NessUUID.(NessUUID.java:204)
> 
> This line in stacktrace show the use of NessUUID, maybe runtime is not in sync with last release ?

Sorry... my mistake, you use bitsy version 3.4.0 (with NessUUID) and NOT 3.4.1 (with standard UUID), so stacktrace is ok.

**Comment by giovibal** [comment]
It works with gremlin-server:3.4.6 instead of gremlin-server:3.4.9, 

```
$ docker-compose run --rm test list
Creating issue01-cannot-list-after-create_test_run ... done
Testing...
[]
$ docker-compose run --rm test create
Creating issue01-cannot-list-after-create_test_run ... done
Testing...
{
  value: Vertex {
    id: '608bb67a-9324-461e-9bc3-7f0c957b8916',
    label: 'Person',
    properties: undefined
  },
  done: false
}
$ docker-compose run --rm test list
Creating issue01-cannot-list-after-create_test_run ... done
Testing...
[
  Map(3) {
    EnumValue { typeName: 'T', elementName: 'id' } => '608bb67a-9324-461e-9bc3-7f0c957b8916',
    EnumValue { typeName: 'T', elementName: 'label' } => 'Person',
    'name' => [ 'Igor' ]
  }
]
```

but i don't know why ...

**Comment by giovibal** [comment]
In BitsyGraphStep.java#76 there is a call to `HasContainer.testAll` and was compiled with gremlin server v3.4.6.
Gremlin server v3.4.6 has a generic implementation that confuse v3.4.9 runtime,
this is the exception rised by gremlin using gremlin console instesad nodejs to test the query:

```
java.lang.NoSuchMethodError: org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer.testAll(Ljava/lang/Object;Ljava/util/List;)Z
        at com.lambdazen.bitsy.gremlin.BitsyGraphStep.lambda$lookupVertices$7(BitsyGraphStep.java:76)
        at org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils$4.advance(IteratorUtils.java:300)
        at org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils$4.hasNext(IteratorUtils.java:269)
        at org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep.processNextStart(GraphStep.java:149)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:145)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.hasNext(ExpandableStepIterator.java:47)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.ReducingBarrierStep.processAllStarts(ReducingBarrierStep.java:82)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.ReducingBarrierStep.processNextStart(ReducingBarrierStep.java:112)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:145)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(ExpandableStepIterator.java:55)
        at org.apache.tinkerpop.gremlin.process.traversal.step.map.MapStep.processNextStart(MapStep.java:36)
        at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:145)
        at org.apache.tinkerpop.gremlin.process.traversal.util.DefaultTraversal.hasNext(DefaultTraversal.java:196)
        at org.apache.tinkerpop.gremlin.server.op.AbstractOpProcessor.handleIterator(AbstractOpProcessor.java:93)
        at org.apache.tinkerpop.gremlin.server.op.AbstractEvalOpProcessor.lambda$evalOpInternal$5(AbstractEvalOpProcessor.java:264)
        at org.apache.tinkerpop.gremlin.groovy.engine.GremlinExecutor.lambda$eval$0(GremlinExecutor.java:283)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

If I recompile bitsy with gremlin v3.4.9 dependency it works fine.

**Comment by giovibal** [comment]
This is my patched version with precompiled `bitsy-3.4.2.jar` that depends on `org.apache.tinkerpop:gremlin-core:3.4.9`
https://github.com/giovibal/bitsy-playground

Please chechout and test if it's ok

**Comment by lambdazen** [comment]
Thanks @giovibal 

If this resolves the issue, please create another PR & I can release 3.4.2.

**Comment by lambdazen** [comment]
Release 3.4.2 includes fix from @giovibal to compile against gremlin 3.4.9.

**Comment by ihoro** [comment]
I've tried bitsy 3.4.2 and my simple create+list test works now.
Thank you all.
